### ISSUE TITLE
aarch64 floating point argument errors

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#1117]
Here is an example program.
```c
$ cat float.c
#include <stdio.h>

float float_add(float a, float b)
{
        fprintf(stderr, "a = %f, b = %f\n", a, b);
        return a + b;
}

double double_add(double a, double b)
{
        fprintf(stderr, "a = %lf, b = %lf\n", a, b);
        return a + b;
}

int main(int argc, char *argv[])
{
        double c;

        c = float_add(-0.1, 0.2);
        fprintf(stderr, "c = %lf\n", c);

        c = double_add(-0.11, 0.22);
        fprintf(stderr, "c = %lf\n", c);

        return c > 0;
}
```
The return values are incorrectly recorded.
```
$ uftrace -a --no-libcall a.out
a = -0.100000, b = 0.200000
c = 0.100000
a = -0.110000, b = 0.220000
c = 0.110000
# DURATION     TID     FUNCTION
            [  4320] | main(0xffff83618d24, 0xffffcfaf9da8) {
  87.031 us [  4320] |   float_add(-0.100000, 0.200000) = -5892472832.000000;
  40.885 us [  4320] |   double_add(-0.110000, 0.220000) = 0.000000;
 206.457 us [  4320] | } = 1; /* main */
```
Unexpectedly, it shows the original program value is also manipulated without `--no-libcall`.
```
$ uftrace -a a.out
a = 1213624708152897478526956949371466656365453977070085167863541782041015626505309452532498268575431296447550982841406775763334082854912.000000, b = 0.000000
c = 0.100000
a = -0.110000, b = 0.220000
c = 0.110000
# DURATION     TID     FUNCTION
   7.240 us [  4447] | __monstartup();
   4.739 us [  4447] | __cxa_atexit();
            [  4447] | main(0xffffb8d21d24, 0xffffdbb1fc18) {
            [  4447] |   float_add(-0.100000, 0.200000) {
   1.463 ms [  4447] |     fprintf(&_IO_2_1_stderr_, "a = %f, b = %f\n") = 159;
   1.473 ms [  4447] |   } = -100192859642134528.000000; /* float_add */
  27.552 us [  4447] |   fprintf(&_IO_2_1_stderr_, "c = %lf\n") = 13;
            [  4447] |   double_add(-0.110000, 0.220000) {
  27.344 us [  4447] |     fprintf(&_IO_2_1_stderr_, "a = %lf, b = %lf\n") = 28;
  32.656 us [  4447] |   } = 0.000000; /* double_add */
  22.656 us [  4447] |   fprintf(&_IO_2_1_stderr_, "c = %lf\n") = 13;
   1.575 ms [  4447] | } = 1; /* main */
```

**Comment by honggyukim** [comment]
The following change fixes the first problem.
```diff
diff --git a/arch/aarch64/mcount-support.c b/arch/aarch64/mcount-support.c
index cd0531d1..398d8f7b 100644
--- a/arch/aarch64/mcount-support.c
+++ b/arch/aarch64/mcount-support.c
@@ -168,22 +168,10 @@ void mcount_arch_get_retval(struct mcount_arg_context *ctx,
                spec->size = 8;

        /* type of return value cannot be FLOAT, so check format instead */
-       if (spec->fmt == ARG_FMT_FLOAT) {
-               long *float_retval = ctx->retval - 2;
-
-               if (spec->size <= 4) {
-                       asm volatile ("ldr s0, %1\n"
-                                     "str s0, %0\n" :
-                                     "=m" (ctx->val.v) : "m" (float_retval));
-               }
-               else {
-                       asm volatile ("ldr d0, %1\n"
-                                     "str d0, %0\n" :
-                                     "=m" (ctx->val.v) : "m" (float_retval));
-               }
-       }
-       else
-               memcpy(ctx->val.v, ctx->retval, spec->size);
+       if (spec->fmt == ARG_FMT_FLOAT)
+               ctx->retval -= 2;
+
+       mcount_memcpy1(ctx->val.v, ctx->retval, spec->size);
 }

 unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
```
```
$ uftrace -a --no-libcall a.out
a = -0.100000, b = 0.200000
c = 0.100000
a = -0.110000, b = 0.220000
c = 0.110000
# DURATION     TID     FUNCTION
            [  4594] | main(0xffff8ee72d24, 0xffffd59341e8) {
  90.833 us [  4594] |   float_add(-0.100000, 0.200000) = 0.100000;
  30.052 us [  4594] |   double_add(-0.110000, 0.220000) = 0.110000;
 188.176 us [  4594] | } = 1; /* main */
```
But the second problem still remains.
```
$ uftrace -a a.out
a = 1213624708152897478526956949371466656365453977070085167863541782041015626505309452532498268575431296447550982841406775763334082854912.000000, b = 0.000000
c = 0.100000
a = -0.110000, b = 0.220000
c = 0.110000
# DURATION     TID     FUNCTION
   6.146 us [  4617] | __monstartup();
   4.010 us [  4617] | __cxa_atexit();
            [  4617] | main(0xffff9b602d24, 0xffffddb3b158) {
            [  4617] |   float_add(-0.100000, 0.200000) {
   1.306 ms [  4617] |     fprintf(&_IO_2_1_stderr_, "a = %f, b = %f\n") = 159;
   1.316 ms [  4617] |   } = 0.100000; /* float_add */
  35.990 us [  4617] |   fprintf(&_IO_2_1_stderr_, "c = %lf\n") = 13;
            [  4617] |   double_add(-0.110000, 0.220000) {
  27.604 us [  4617] |     fprintf(&_IO_2_1_stderr_, "a = %lf, b = %lf\n") = 28;
  33.125 us [  4617] |   } = 0.110000; /* double_add */
  23.386 us [  4617] |   fprintf(&_IO_2_1_stderr_, "c = %lf\n") = 13;
   1.425 ms [  4617] | } = 1; /* main */
```

**Comment by honggyukim** [comment]
The following doesn't make a problem.
```
$ uftrace -F float_add -a -T fprintf@arg1/p,arg2,retval a.out
a = -0.100000, b = 0.200000
c = 0.100000
a = -0.110000, b = 0.220000
c = 0.110000
# DURATION     TID     FUNCTION
            [  7704] | float_add(-0.100000, 0.200000) {
  84.947 us [  7704] |   fprintf(&_IO_2_1_stderr_, 0xaaaab3f07b08) = 28;
 101.457 us [  7704] | } = 0.100000; /* float_add */
```
But the following with `arg2/s` makes the same problem.
```
$ uftrace -F float_add -a -T fprintf@arg1/p,arg2/s,retval a.out
a = 1213624708152897478526956949371466656365453977070085167863541782041015626505309452532498268575431296447550982841406775763334082854912.000000, b = 0.000000
c = 0.100000
a = -0.110000, b = 0.220000
c = 0.110000
# DURATION     TID     FUNCTION
            [  7721] | float_add(-0.100000, 0.200000) {
   1.359 ms [  7721] |   fprintf(&_IO_2_1_stderr_, "a = %f, b = %f\n") = 159;
   1.376 ms [  7721] | } = 0.100000; /* float_add */
```

**Comment by honggyukim** [comment]
The above problem can be fixed with the following change as well.
```diff
diff --git a/arch/aarch64/mcount-arch.h b/arch/aarch64/mcount-arch.h
index 7afe6552..69efe521 100644
--- a/arch/aarch64/mcount-arch.h
+++ b/arch/aarch64/mcount-arch.h
@@ -26,7 +26,9 @@ struct mcount_regs {
 #define ARCH_MAX_REG_ARGS  8
 #define ARCH_MAX_FLOAT_REGS  8

+#define HAVE_MCOUNT_ARCH_CONTEXT
 struct mcount_arch_context {
+       double d[ARCH_MAX_FLOAT_REGS];
 };

 #define ARCH_PLT0_SIZE  32
@@ -193,3 +181,27 @@ unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
        sym = &pd->dsymtab.sym[0];
        return sym->addr - ARCH_PLT0_SIZE;
 }
+
+void mcount_save_arch_context(struct mcount_arch_context *ctx)
+{
+       asm volatile ("str d0, %0\n" : "=m" (ctx->d[0]));
+       asm volatile ("str d1, %0\n" : "=m" (ctx->d[1]));
+       asm volatile ("str d2, %0\n" : "=m" (ctx->d[2]));
+       asm volatile ("str d3, %0\n" : "=m" (ctx->d[3]));
+       asm volatile ("str d4, %0\n" : "=m" (ctx->d[4]));
+       asm volatile ("str d5, %0\n" : "=m" (ctx->d[5]));
+       asm volatile ("str d6, %0\n" : "=m" (ctx->d[6]));
+       asm volatile ("str d7, %0\n" : "=m" (ctx->d[7]));
+}
+
+void mcount_restore_arch_context(struct mcount_arch_context *ctx)
+{
+       asm volatile ("ldr d0, %0\n" :: "m" (ctx->d[0]));
+       asm volatile ("ldr d1, %0\n" :: "m" (ctx->d[1]));
+       asm volatile ("ldr d2, %0\n" :: "m" (ctx->d[2]));
+       asm volatile ("ldr d3, %0\n" :: "m" (ctx->d[3]));
+       asm volatile ("ldr d4, %0\n" :: "m" (ctx->d[4]));
+       asm volatile ("ldr d5, %0\n" :: "m" (ctx->d[5]));
+       asm volatile ("ldr d6, %0\n" :: "m" (ctx->d[6]));
+       asm volatile ("ldr d7, %0\n" :: "m" (ctx->d[7]));
+}
```
The result shows that replay output is correct now.
```
$ uftrace -F float_add -a -T fprintf@arg1/p,arg2/s,retval a.out
a = -0.100000, b = 0.200000
c = 0.100000
a = -0.110000, b = 0.220000
c = 0.110000
# DURATION     TID     FUNCTION
            [  9603] | float_add(-0.100000, 0.200000) {
   1.193 ms [  9603] |   fprintf(&_IO_2_1_stderr_, "a = %f, b = %f\n") = 28;
   1.208 ms [  9603] | } = 0.100000; /* float_add */
```
The full replay output is also correct.
```
$ uftrace -a a.out
a = -0.100000, b = 0.200000
c = 0.100000
a = -0.110000, b = 0.220000
c = 0.110000
# DURATION     TID     FUNCTION
   7.552 us [  9595] | __monstartup();
   4.167 us [  9595] | __cxa_atexit();
            [  9595] | main(0xffff8ef75d24, 0xffffd78c7298) {
            [  9595] |   float_add(-0.100000, 0.200000) {
   1.221 ms [  9595] |     fprintf(&_IO_2_1_stderr_, "a = %f, b = %f\n") = 28;
   1.229 ms [  9595] |   } = 0.100000; /* float_add */
  33.125 us [  9595] |   fprintf(&_IO_2_1_stderr_, "c = %lf\n") = 13;
            [  9595] |   double_add(-0.110000, 0.220000) {
  26.979 us [  9595] |     fprintf(&_IO_2_1_stderr_, "a = %lf, b = %lf\n") = 28;
  32.551 us [  9595] |   } = 0.110000; /* double_add */
  22.447 us [  9595] |   fprintf(&_IO_2_1_stderr_, "c = %lf\n") = 13;
   1.336 ms [  9595] | } = 1; /* main */
```

**Comment by honggyukim** [comment]
@namhyung If you agree on both changes, then I will send a PR.

**Comment by namhyung** [comment]
Hmm.. I don't know why the inline asm doesn't work but if it works all platforms I think it'd better using it as it's simpler.

Also for the context save/restore, does it work if you use `-march=armv8-a+nofp` without your patch?

**Comment by honggyukim** [comment]
I didn't change it with `memcpy` in the end. Please refer to my commit in PR #1122 - https://github.com/namhyung/uftrace/pull/1122/commits/251ba74a7283664b330649c239dfea20dd8f9dae.

I also tried to recompile with `-march=armv8-a+nofp`, but shows some compilation errors.
```
$ make
  CC       uftrace.o
  CC       cmds/script.o
  CC       cmds/replay.o
In file included from /home/honggyu/work/uftrace/utils/symbol.h:15,
                 from /home/honggyu/work/uftrace/uftrace.h:11,
                 from /home/honggyu/work/uftrace/cmds/replay.c:8:
/home/honggyu/work/uftrace/cmds/replay.c: In function ‘print_event’:
/home/honggyu/work/uftrace/utils/utils.h:162:31: error: ‘+nofp’ feature modifier is incompatible with floating-point argument
 #define pr_color(c, fmt, ...) __pr_color(c,                  fmt, ## __VA_ARGS__)
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/honggyu/work/uftrace/cmds/replay.c:353:4: note: in expansion of macro ‘pr_color’
    pr_color(color, "%s (cycle=%+"PRId64", instructions=%+"PRId64", IPC=%.2f)",
    ^~~~~~~~
```

**Comment by namhyung** [comment]
Hmm.. it seems we need to use it only for the libmcount.

**Comment by honggyukim** [comment]
It seems we need more modification to compile it with `-march=armv8-a+nofp`.
```
$ make
  CC       uftrace.o
  CC       cmds/script.o
  LINK     uftrace
  CC FPIC  libmcount/debug.op
/home/honggyu/work/uftrace/utils/debug.c: In function ‘print_diff_percent’:
/home/honggyu/work/uftrace/utils/debug.c:325:6: error: ‘+nofp’ feature modifier is incompatible with floating-point code
 void print_diff_percent(uint64_t base_nsec, uint64_t pair_nsec)
      ^~~~~~~~~~~~~~~~~~
/home/honggyu/work/uftrace/utils/debug.c:325: confused by earlier errors, bailing out
make: *** [Makefile:228: /home/honggyu/work/uftrace/libmcount/debug.op] Error 1
```
Here is the change I used.
```diff
diff --git a/Makefile b/Makefile
index 68489028..298c7b02 100644
--- a/Makefile
+++ b/Makefile
@@ -75,7 +75,7 @@ DEMANGLER_CFLAGS   = $(COMMON_CFLAGS) $(CFLAGS_$@) $(CFLAGS_demangler)
 SYMBOLS_CFLAGS     = $(COMMON_CFLAGS) $(CFLAGS_$@) $(CFLAGS_symbols)
 TRACEEVENT_CFLAGS  = $(COMMON_CFLAGS) $(CFLAGS_$@) $(CFLAGS_traceevent)
 LIB_CFLAGS         = $(COMMON_CFLAGS) $(CFLAGS_$@) $(CFLAGS_lib)
-LIB_CFLAGS        += -fPIC -fvisibility=hidden -fno-omit-frame-pointer
+LIB_CFLAGS        += -fPIC -fvisibility=hidden -fno-omit-frame-pointer -march='armv8-a+nofp'
 TEST_CFLAGS        = $(COMMON_CFLAGS) -DUNIT_TEST
 
 UFTRACE_LDFLAGS    = $(COMMON_LDFLAGS) $(LDFLAGS_$@) $(LDFLAGS_uftrace)
```

**Comment by honggyukim** [comment]
I tried to recompile it with `-march=armv8-a+nofp` manually modifying `utils/debug.c`, but the result has the same problem.
```
$ uftrace -a a.out
float_add(1213624708152897478526956949371466656365453977070085167863541782041015626505309452532498268575431296447550982841406775763334082854912.000000, 0.000000) = 0.100000
double_add(-0.110000, 0.220000) = 0.110000
# DURATION     TID     FUNCTION
   6.406 us [   910] | __monstartup();
   3.854 us [   910] | __cxa_atexit();
            [   910] | main(0xffffa4532d24, 0xffffcbc04448) {
            [   910] |   float_add(-0.100000, 0.200000) {
   1.401 ms [   910] |     fprintf(&_IO_2_1_stderr_, "float_add(%f, %f)") = 161;
   1.410 ms [   910] |   } = -25199936.000000; /* float_add */
  34.584 us [   910] |   fprintf(&_IO_2_1_stderr_, " = %lf\n") = 12;
            [   910] |   double_add(-0.110000, 0.220000) {
  26.093 us [   910] |     fprintf(&_IO_2_1_stderr_, "double_add(%lf, %lf)") = 31;
  31.667 us [   910] |   } = 0.000000; /* double_add */
  35.989 us [   910] |   fprintf(&_IO_2_1_stderr_, " = %lf\n") = 12;
   1.532 ms [   910] | } = 1; /* main */
```

**Comment by honggyukim** [comment]
Could you please take a look at the change in #1122?

**Comment by namhyung** [comment]
Sorry for the late reply, will do.

**Comment by honggyukim** [comment]
Fixed by a0fbee4 and 251ba74 in #1122.
